libazureinit/src/media.rs

```diff
@@ -66,8 +67,49 @@ fn default_preprov_type() -> String {
     "None".to_owned()
 }
 
-pub fn make_temp_directory() -> Result<(), Box<dyn std::error::Error>> {
-    let file_path = "/run/azure-init/tmp/";
+const PATH_MEDIA_MOUNT: &str = "/run/azure-init/media/";
+
+pub fn mount_media() -> Result<(), Error> {
+    let mount_status = Command::new("mount")
+        .arg("-o")
+        .arg("ro")
+        .arg("/dev/sr0")
+        .arg(PATH_MEDIA_MOUNT)
```

Comment:
We have to create the folder PATH_MEDIA_MOUNT before mounting it right? When creating the folder, we should also restrict the read/write access to root only

Reply:
If you want to ensure it's impossible to mis-use this API (forgetting to call `make_temp_directory`, for example) you can use a pattern (which I saw in [Rust for Rustaceans](https://rust-for-rustaceans.com/)) which creates a little state machine and hides functions until the object is in the right state. For example:

```
// Some zero-sized structs that just provide states for our state machine
struct Mounted;
struct Unmounted;

struct Media<State = Unmounted> {
    mount_path: PathBuf,
    device_path: PathBuf,
    state: std::marker::PhantomData<State>,
}

impl Media<Unmounted> {
    pub fn new(device_path: PathBuf, mount_path: PathBuf) -> Media<Unmounted> {
        Media {
            mount_path,
            device_path,
            state: std::marker::PhantomData,
        }
    }

    // The only thing you can do with the Media struct is call mount which, if it succeeds,
    // moves it into the Mounted state. From there the only thing you can do with it is
    // call unmount(), but you could make other functionality available by adding to the
    // impl Media<Mounted> block.
    pub fn mount(self) -> Result<Media<Mounted>, Error> {
        create_dir_all(&self.mount_path)?;

        let metadata = fs::metadata(&self.mount_path)?;
        let permissions = metadata.permissions();
        let mut new_permissions = permissions.clone();
        new_permissions.set_mode(0o700);
        fs::set_permissions(&self.mount_path, new_permissions)?;

        let mount_status = Command::new("mount")
            .arg("-o")
            .arg("ro")
            .arg("/dev/sr0")
            .arg(PATH_MEDIA_MOUNT)
            .status()?;

        if !mount_status.success() {
            Err(Error::SubprocessFailed {
                command: "mount".to_string(),
                status: mount_status,
            })
        } else {
            Ok(Media {
                mount_path: self.mount_path,
                device_path: self.device_path,
                state: std::marker::PhantomData,
            })
        }
    }
}

impl Media<Mounted> {
    pub fn unmount(self) -> Result<(), Error> {
        let umount_status =
            Command::new("umount").arg(self.mount_path).status()?;
        if !umount_status.success() {
            return Err(Error::SubprocessFailed {
                command: "umount".to_string(),
                status: umount_status,
            });
        }

        let eject_status =
            Command::new("eject").arg(self.device_path).status()?;
        if !eject_status.success() {
            Err(Error::SubprocessFailed {
                command: "eject".to_string(),
                status: eject_status,
            })
        } else {
            Ok(())
        }
    }
}

```

Reply:
@jeremycline Thanks. Rewrote the mount part as you suggested.